<a href="https://colab.research.google.com/github/LuppeCh/F1EDA/blob/EDAampliado1/Exploraci%C3%B3n_de_Datos_Trabajo_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
install.packages("dplyr")
install.packages("ggplot2")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message in install.packages("ggplot2"):
“installation of package ‘ggplot2’ had non-zero exit status”


In [ ]:
library(dplyr)
library(ggplot2)

In [ ]:
wil <- read.csv("williams.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(wil)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        0      0       0        1       
2 Alpine Interlagos 4294        0        0      0       1        1       
3 Alpine Interlagos 4294        0        0      0       2        1       
4 Alpine Interlagos 4294        0        0      0       3        1       
5 Alpine Interlagos 4294        0        0      0       4        1       
6 Alpine Interlagos 4294        0        0      0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 0          0.513507     ⋯ 0               0             0           0        
2 0          1.449180     ⋯ 0               0             0           0        
3 0          2.384850     ⋯ 0               0             0           0        
4 0          3.320510     ⋯ 0               0             0           0        
5 0          4.256180     ⋯ 0               0             0           0        
6 0          5.194280     ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
wil$carId <- "WIL"

In [ ]:
tail(wil)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
25759 WIL   Interlagos 4294        5        5      0       4288     0       
25760 WIL   Interlagos 4294        5        5      0       4289     0       
25761 WIL   Interlagos 4294        5        5      0       4290     0       
25762 WIL   Interlagos 4294        5        5      0       4291     0       
25763 WIL   Interlagos 4294        5        5      0       4292     0       
25764 WIL   Interlagos 4294        5        5      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
25759 -1         -1           ⋯ -1              -1            -1         
25760 -1         -1           ⋯ -1              -1            -1         
25761 -1         -1           ⋯ -1              -1            -1         
25762 -1         -1           ⋯ -1              -1            -1         
25763 -1         -1           ⋯ -1              -1            -1         
25764 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
25759 -1        -1        -1         -1          -1        -1          
25760 -1        -1        -1         -1          -1        -1          
25761 -1        -1        -1         -1          -1        -1          
25762 -1        -1        -1         -1          -1        -1          
25763 -1        -1        -1         -1          -1        -1          
25764 -1        -1        -1         -1          -1        -1          
      engine_seized
25759 -1           
25760 -1           
25761 -1           
25762 -1           
25763 -1           
25764 -1

Observamos que lapNum tiene un valor mayor al esperado (4), ya que arranca a contabilizar las vueltas desde 0, por lo que el lapNum = 5 corresponderia a la vuelta 6. Además observamos que los valores de muchas variables toman el valor "-1", lo cual es inconsistente en algunas variables como "lap_distance", por lo que aunque debemos eliminar todos los registros de esta vuelta, vamos a realizar un breve análisis que nos permita comprender que sucedió.

In [ ]:
wil <- wil %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

Para trabajar más cómodos, decidimos eliminar todas aquellas columnas que no vamos a utilizar ni ahora ni a futuro.

In [ ]:
inconsistencia <- wil %>%

  select(where(is.numeric)) %>%

  summarise(
    across(
      everything(),
      ~ sum(. == -1, na.rm = TRUE)
    )
  )

inconsistencia

lapNum lap_time lap_distance lap_time_invalid binIndex velocity_X velocity_Y
1 0      3237     3237         3237             0        3237       3237      
  velocity_Z
1 3237

Creamos un pequeño dataset que nos permite ver cuantas de las variables númericas existentes en nuestro dataset tienen un valor igual a "-1". Al imprimirla observamos que 6 de las 8 variables tienen 3237 registros de este tipo. Sospechamos que cuando una de estas variables toma el valor -1, las otras 5 también.

In [ ]:
sospechosas <- c(
  "lap_time",
  "lap_distance",
  "lap_time_invalid",
  "velocity_X",
  "velocity_Y",
  "velocity_Z"
)


conteo_sospechosas <- wil %>%
  summarise(
    cant_sospechosas = sum(if_all(all_of(sospechosas), ~ . == -1), na.rm = TRUE)
  ) %>%
  pull()

conteo_sospechosas

[1] 3237

Se confirma nuestra teoría. Además, por lo leído en el archivo de especificaciones de EA SPORTS F1, este valor suele significar que no se ha cruzado la línea de meta, lo que puede indicar que estos valores pertenezcan a:

1. Una vuelta de calentamiento.
2. Los metros que sigue recorriendo el automóvil antes de frenar, luego de cruzar la meta.

In [ ]:
final_valido <- wil %>%
  filter(lapNum == 5) %>%
  group_by(lap_time_invalid) %>%
  summarise(
    N_Registros = n(),
    .groups = 'drop'
  )

  final_valido

lap_time_invalid N_Registros
1 -1               3237       
2  0               1057

Confirmamos que todos estos valores pertenecen a la vuelta 6 (lapNum = 5), lo que indicaria que todos estos registros se deben a la razón 2.

In [ ]:
wil <- wil %>%
  filter(lapNum < 5)

Borramos esta vuelta.

In [ ]:
mcl <- read.csv("mcclaren.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(mcl)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        21     0       0        1       
2 Alpine Interlagos 4294        0        21     0       1        1       
3 Alpine Interlagos 4294        0        21     0       2        1       
4 Alpine Interlagos 4294        0        21     0       3        1       
5 Alpine Interlagos 4294        0        21     0       4        1       
6 Alpine Interlagos 4294        0        21     0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 21         0.40625      ⋯ 0               0             0           0        
2 21         1.41602      ⋯ 0               0             0           0        
3 21         2.42578      ⋯ 0               0             0           0        
4 21         3.43555      ⋯ 0               0             0           0        
5 21         4.44531      ⋯ 0               0             0           0        
6 21         5.45703      ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
mcl$carId <- "MCL"

In [ ]:
mcl$lapNum <- mcl$lapNum - 21

In [ ]:
tail(mcl)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
25759 MCL   Interlagos 4294        5        5      0       4288     0       
25760 MCL   Interlagos 4294        5        5      0       4289     0       
25761 MCL   Interlagos 4294        5        5      0       4290     0       
25762 MCL   Interlagos 4294        5        5      0       4291     0       
25763 MCL   Interlagos 4294        5        5      0       4292     0       
25764 MCL   Interlagos 4294        5        5      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
25759 -1         -1           ⋯ -1              -1            -1         
25760 -1         -1           ⋯ -1              -1            -1         
25761 -1         -1           ⋯ -1              -1            -1         
25762 -1         -1           ⋯ -1              -1            -1         
25763 -1         -1           ⋯ -1              -1            -1         
25764 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
25759 -1        -1        -1         -1          -1        -1          
25760 -1        -1        -1         -1          -1        -1          
25761 -1        -1        -1         -1          -1        -1          
25762 -1        -1        -1         -1          -1        -1          
25763 -1        -1        -1         -1          -1        -1          
25764 -1        -1        -1         -1          -1        -1          
      engine_seized
25759 -1           
25760 -1           
25761 -1           
25762 -1           
25763 -1           
25764 -1

In [ ]:
mcl <- mcl %>%
  filter(lapNum < 5)

In [ ]:
mcl <- mcl %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

In [ ]:
red <- read.csv("redbull.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(red)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        16     0       0        1       
2 Alpine Interlagos 4294        0        16     0       1        1       
3 Alpine Interlagos 4294        0        16     0       2        1       
4 Alpine Interlagos 4294        0        16     0       3        1       
5 Alpine Interlagos 4294        0        16     0       4        1       
6 Alpine Interlagos 4294        0        16     0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 16         2.81250      ⋯ 0               0             0           0        
2 16         2.81250      ⋯ 0               0             0           0        
3 16         2.81250      ⋯ 0               0             0           0        
4 16         3.82617      ⋯ 0               0             0           0        
5 16         4.83984      ⋯ 0               0             0           0        
6 16         5.85352      ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
red$carId <-"RED"

In [ ]:
red$lapNum <- red$lapNum - 16

In [ ]:
tail(red)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
25759 RED   Interlagos 4294        5        5      0       4288     0       
25760 RED   Interlagos 4294        5        5      0       4289     0       
25761 RED   Interlagos 4294        5        5      0       4290     0       
25762 RED   Interlagos 4294        5        5      0       4291     0       
25763 RED   Interlagos 4294        5        5      0       4292     0       
25764 RED   Interlagos 4294        5        5      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
25759 -1         -1           ⋯ -1              -1            -1         
25760 -1         -1           ⋯ -1              -1            -1         
25761 -1         -1           ⋯ -1              -1            -1         
25762 -1         -1           ⋯ -1              -1            -1         
25763 -1         -1           ⋯ -1              -1            -1         
25764 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
25759 -1        -1        -1         -1          -1        -1          
25760 -1        -1        -1         -1          -1        -1          
25761 -1        -1        -1         -1          -1        -1          
25762 -1        -1        -1         -1          -1        -1          
25763 -1        -1        -1         -1          -1        -1          
25764 -1        -1        -1         -1          -1        -1          
      engine_seized
25759 -1           
25760 -1           
25761 -1           
25762 -1           
25763 -1           
25764 -1

In [ ]:
red <- red %>%
  filter(lapNum < 5)

In [ ]:
red <- red %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

In [ ]:
alp <- read.csv("alpine.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(alp)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        10     0       0        1       
2 Alpine Interlagos 4294        0        10     0       1        1       
3 Alpine Interlagos 4294        0        10     0       2        1       
4 Alpine Interlagos 4294        0        10     0       3        1       
5 Alpine Interlagos 4294        0        10     0       4        1       
6 Alpine Interlagos 4294        0        10     0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 10         1.63281      ⋯ 0               0             0           0        
2 10         1.63281      ⋯ 0               0             0           0        
3 10         2.62500      ⋯ 0               0             0           0        
4 10         3.61719      ⋯ 0               0             0           0        
5 10         4.60938      ⋯ 0               0             0           0        
6 10         5.60156      ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
alp$carId <- "ALP"

In [ ]:
alp$lapNum <- alp$lapNum - 10

In [ ]:
tail(alp)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
30053 ALP   Interlagos 4294        6        6      0       4288     0       
30054 ALP   Interlagos 4294        6        6      0       4289     0       
30055 ALP   Interlagos 4294        6        6      0       4290     0       
30056 ALP   Interlagos 4294        6        6      0       4291     0       
30057 ALP   Interlagos 4294        6        6      0       4292     0       
30058 ALP   Interlagos 4294        6        6      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
30053 -1         -1           ⋯ -1              -1            -1         
30054 -1         -1           ⋯ -1              -1            -1         
30055 -1         -1           ⋯ -1              -1            -1         
30056 -1         -1           ⋯ -1              -1            -1         
30057 -1         -1           ⋯ -1              -1            -1         
30058 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
30053 -1        -1        -1         -1          -1        -1          
30054 -1        -1        -1         -1          -1        -1          
30055 -1        -1        -1         -1          -1        -1          
30056 -1        -1        -1         -1          -1        -1          
30057 -1        -1        -1         -1          -1        -1          
30058 -1        -1        -1         -1          -1        -1          
      engine_seized
30053 -1           
30054 -1           
30055 -1           
30056 -1           
30057 -1           
30058 -1

In [ ]:
alp <- alp %>%
  filter(lapNum < 5)

In [ ]:
alp <- alp %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

In [ ]:
mer <- read.csv("mercedes.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(mer)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        5      0       0        1       
2 Alpine Interlagos 4294        0        5      0       1        1       
3 Alpine Interlagos 4294        0        5      0       2        1       
4 Alpine Interlagos 4294        0        5      0       3        1       
5 Alpine Interlagos 4294        0        5      0       4        1       
6 Alpine Interlagos 4294        0        5      0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 5          3.34570      ⋯ 0               0             0           0        
2 5          3.34570      ⋯ 0               0             0           0        
3 5          3.34570      ⋯ 0               0             0           0        
4 5          3.34570      ⋯ 0               0             0           0        
5 5          4.41357      ⋯ 0               0             0           0        
6 5          5.48145      ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
mer$carId <-"MER"

In [ ]:
mer$lapNum <- mer$lapNum - 5

In [ ]:
tail(mer)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
25759 MER   Interlagos 4294        5        5      0       4288     0       
25760 MER   Interlagos 4294        5        5      0       4289     0       
25761 MER   Interlagos 4294        5        5      0       4290     0       
25762 MER   Interlagos 4294        5        5      0       4291     0       
25763 MER   Interlagos 4294        5        5      0       4292     0       
25764 MER   Interlagos 4294        5        5      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
25759 -1         -1           ⋯ -1              -1            -1         
25760 -1         -1           ⋯ -1              -1            -1         
25761 -1         -1           ⋯ -1              -1            -1         
25762 -1         -1           ⋯ -1              -1            -1         
25763 -1         -1           ⋯ -1              -1            -1         
25764 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
25759 -1        -1        -1         -1          -1        -1          
25760 -1        -1        -1         -1          -1        -1          
25761 -1        -1        -1         -1          -1        -1          
25762 -1        -1        -1         -1          -1        -1          
25763 -1        -1        -1         -1          -1        -1          
25764 -1        -1        -1         -1          -1        -1          
      engine_seized
25759 -1           
25760 -1           
25761 -1           
25762 -1           
25763 -1           
25764 -1

In [ ]:
mer <- mer %>%
  filter(lapNum < 5)

In [ ]:
mer <- mer %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

In [ ]:
mer <- mer %>%
  mutate(
    sector = case_when(
      binIndex >= 0 & binIndex <= 1430 ~ "S1",
      binIndex >= 1431 & binIndex <= 2861 ~ "S2",
      binIndex >= 2862 & binIndex <= 4293 ~ "S3",
    ))

In [ ]:
fer <- read.csv("ferrari.csv", header = TRUE, dec = ".", sep = "\t")

In [ ]:
head(fer)

carId  trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
1 Alpine Interlagos 4294        0        31     0       0        1       
2 Alpine Interlagos 4294        0        31     0       1        1       
3 Alpine Interlagos 4294        0        31     0       2        1       
4 Alpine Interlagos 4294        0        31     0       3        1       
5 Alpine Interlagos 4294        0        31     0       4        1       
6 Alpine Interlagos 4294        0        31     0       5        1       
  lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage es_damage
1 31         0.93750      ⋯ 0               0             0           0        
2 31         1.80313      ⋯ 0               0             0           0        
3 31         2.66875      ⋯ 0               0             0           0        
4 31         3.53438      ⋯ 0               0             0           0        
5 31         4.40000      ⋯ 0               0             0           0        
6 31         5.26562      ⋯ 0               0             0           0        
  ce_damage ice_damage mguk_damage tc_damage engine_blown engine_seized
1 0         0          0           0         0            0            
2 0         0          0           0         0            0            
3 0         0          0           0         0            0            
4 0         0          0           0         0            0            
5 0         0          0           0         0            0            
6 0         0          0           0         0            0

In [ ]:
fer$carId <- "FER"

In [ ]:
fer$lapNum <- fer$lapNum - 31

In [ ]:
tail(fer)

carId trackId    trackLength lapIndex lapNum lapFlag binIndex validBin
21465 FER   Interlagos 4294        4        4      0       4288     0       
21466 FER   Interlagos 4294        4        4      0       4289     0       
21467 FER   Interlagos 4294        4        4      0       4290     0       
21468 FER   Interlagos 4294        4        4      0       4291     0       
21469 FER   Interlagos 4294        4        4      0       4292     0       
21470 FER   Interlagos 4294        4        4      0       4293     0       
      lap_number lap_distance ⋯ gear_box_damage engine_damage mguh_damage
21465 -1         -1           ⋯ -1              -1            -1         
21466 -1         -1           ⋯ -1              -1            -1         
21467 -1         -1           ⋯ -1              -1            -1         
21468 -1         -1           ⋯ -1              -1            -1         
21469 -1         -1           ⋯ -1              -1            -1         
21470 -1         -1           ⋯ -1              -1            -1         
      es_damage ce_damage ice_damage mguk_damage tc_damage engine_blown
21465 -1        -1        -1         -1          -1        -1          
21466 -1        -1        -1         -1          -1        -1          
21467 -1        -1        -1         -1          -1        -1          
21468 -1        -1        -1         -1          -1        -1          
21469 -1        -1        -1         -1          -1        -1          
21470 -1        -1        -1         -1          -1        -1          
      engine_seized
21465 -1           
21466 -1           
21467 -1           
21468 -1           
21469 -1           
21470 -1

En este caso si tenemos la cantidad correcta de vueltas, pero como podemos ver hay muchisimos valores "-1", por lo que debemos borrar esta vuelta y generarla de manera artificial.

In [ ]:
fer <- fer %>%
  filter(lapNum < 4)

In [ ]:
fer <- fer %>%
  select(
    carId,
    lapNum,
    lap_time,
    lap_distance,
    lap_time_invalid,
    binIndex,
    velocity_X,
    velocity_Y,
    velocity_Z
  )

In [ ]:
vueltas_anuladas_fer <- fer %>%
  group_by(lapNum) %>%
  summarise(lap_time_invalid = max(lap_time_invalid))

Para generar de forma artificial la quinta vuelta, debemos realizarlo promediando los valores de las vueltas validas, por lo que visualizamos cuales son (las que tienen lap_time_valid = 0)

In [ ]:
vueltas_anuladas_fer

lapNum lap_time_invalid
1 0      0               
2 1      1               
3 2      1               
4 3      0

Solamente podemos trabajar con la vuelta 1 y 4.

In [ ]:
vueltas_validas <- fer %>%
  filter(lapNum %in% c(0, 3))

In [ ]:
vuelta_5 <- vueltas_validas %>%
  group_by(binIndex) %>%
  summarise(
    lap_distance = mean(lap_distance, na.rm = TRUE),
    lap_time = mean(lap_time, na.rm = TRUE),
    velocity_X = mean(velocity_X, na.rm = TRUE),
    velocity_Y = mean(velocity_Y, na.rm = TRUE),
    velocity_Z = mean(velocity_Z, na.rm = TRUE),

    carId = "FER",
    lap_time_invalid = 0
  ) %>%

  ungroup()%>%

  mutate(
    lapNum = 4
  )

Las filtramos en el dataset "vueltas_validas" y construimos la "vuelta_5" calculando los promedios en la variables indicadas, despues le pasamos el valor del id, que la vuelta es valida y que es la número 5.

In [ ]:
fer <- bind_rows(fer, vuelta_5)

La agregamos a nuestro dataset de ferrari.

In [ ]:
telemetria <- bind_rows(fer, red, mcl, mer, alp, wil)

Únimos los 6 datasets en uno único.

In [ ]:
telemetria <- telemetria %>%
  mutate(
    sector = case_when(
      binIndex >= 0 & binIndex <= 1430 ~ "S1",
      binIndex >= 1431 & binIndex <= 2861 ~ "S2",
      binIndex >= 2862 & binIndex <= 4293 ~ "S3",
    ),
    track_lenght = 4294
    )

Creamos la columna sector, indicando el valor que toma según el binIndex.

In [ ]:
telemetria <- telemetria %>%
  mutate(
    velocity_total_ms = sqrt(
      velocity_X^2 +
      velocity_Y^2 +
      velocity_Z^2
    ),

    velocidad = velocity_total_ms * 3.6
  )

telemetria <- telemetria %>%
  select(-velocity_total_ms)

In [ ]:
inconsistencia_global <- telemetria %>%

  select(where(is.numeric)) %>%

  summarise(
    across(
      everything(),
      ~ sum(. == -1, na.rm = TRUE)
    )
  )

inconsistencia_global

lapNum lap_time lap_distance lap_time_invalid binIndex velocity_X velocity_Y
1 0      0        0            0                0        0          0         
  velocity_Z track_lenght velocidad
1 0          0            0

In [ ]:
telemetria <- telemetria %>%
  mutate(lap_number = lapNum + 1)

In [ ]:
unique_invalid <- unique(telemetria$lap_time_invalid)
unique_invalid

[1] 0 1

In [ ]:
porcentaje_invalido <- telemetria %>%
  group_by(carId, lapNum) %>%
  summarise(
    total_registros = n(),
    obs_invalidas = sum(lap_time_invalid == 1, na.rm = TRUE),
    porcentaje_invalido = (obs_invalidas / total_registros) * 100
  ) %>%
  ungroup() %>%
  arrange(desc(porcentaje_invalido), carId, lapNum)

porcentaje_invalido

`summarise()` has grouped output by 'carId'. You can override using the
`.groups` argument.


carId lapNum total_registros obs_invalidas porcentaje_invalido
1  ALP   0      4294            3746          87.23801           
2  RED   1      4294            3658          85.18864           
3  MCL   4      4294            3544          82.53377           
4  RED   0      4294            2791          64.99767           
5  ALP   1      4294            2541          59.17559           
6  RED   3      4294            2241          52.18910           
7  WIL   4      4294            2044          47.60130           
8  WIL   1      4294            2042          47.55473           
9  WIL   3      4294            2033          47.34513           
10 FER   1      4294            2027          47.20540           
11 FER   2      4294            2026          47.18211           
12 ALP   4      4294            1466          34.14066           
13 ALP   2      4294               0           0.00000           
14 ALP   3      4294               0           0.00000           
15 FER   0      4294               0           0.00000           
16 FER   3      4294               0           0.00000           
17 FER   4      4294               0           0.00000           
18 MCL   0      4294               0           0.00000           
19 MCL   1      4294               0           0.00000           
20 MCL   2      4294               0           0.00000           
21 MCL   3      4294               0           0.00000           
22 MER   0      4294               0           0.00000           
23 MER   1      4294               0           0.00000           
24 MER   2      4294               0           0.00000           
25 MER   3      4294               0           0.00000           
26 MER   4      4294               0           0.00000           
27 RED   2      4294               0           0.00000           
28 RED   4      4294               0           0.00000           
29 WIL   0      4294               0           0.00000           
30 WIL   2      4294               0           0.00000

In [ ]:
vueltas_a_invalidar <- c(
  "ALP_0", "RED_1", "MCL_4", "RED_0", "ALP_1", "RED_3",
  "WIL_4", "WIL_1", "WIL_3", "FER_1", "FER_2", "ALP_4")

telemetria_vueltas_invalidadas <- telemetria %>%
  mutate(lap_id = paste(carId, lapNum, sep = "_")) %>%
  mutate(
    lap_time_invalid = if_else(lap_id %in% vueltas_a_invalidar, 1, lap_time_invalid)) %>%
  select(-lap_id)

In [ ]:
registros <- nrow(telemetria_vueltas_invalidadas)
print(paste("cant. registros:", registros))

escuderias <- unique(telemetria_vueltas_invalidadas$carId)
print(escuderias)

cant_lap <- unique(telemetria_vueltas_invalidadas$lapNum)
print(cant_lap)

cant_lap <- unique(telemetria_vueltas_invalidadas$lap_number)
print(cant_lap)

any(is.na(telemetria_vueltas_invalidadas))

[1] "cant. registros: 128820"
[1] "FER" "RED" "MCL" "MER" "ALP" "WIL"
[1] 0 1 2 3 4
[1] 1 2 3 4 5


[1] FALSE

In [ ]:
summary(telemetria_vueltas_invalidadas)

    carId               lapNum     lap_time      lap_distance      
 Length:128820      Min.   :0   Min.   : 0.00   Min.   :   0.0312  
 Class :character   1st Qu.:1   1st Qu.:17.39   1st Qu.:1073.5850  
 Mode  :character   Median :2   Median :34.08   Median :2146.9850  
                    Mean   :2   Mean   :37.37   Mean   :2146.9966  
                    3rd Qu.:3   3rd Qu.:58.95   3rd Qu.:3220.5300  
                    Max.   :4   Max.   :83.37   Max.   :4294.0000  
 lap_time_invalid    binIndex      velocity_X       velocity_Y     
 Min.   :0.0      Min.   :   0   Min.   : 3.316   Min.   :-9.7585  
 1st Qu.:0.0      1st Qu.:1073   1st Qu.:48.901   1st Qu.:-0.1395  
 Median :0.0      Median :2146   Median :65.286   Median : 0.2474  
 Mean   :0.4      Mean   :2146   Mean   :62.746   Mean   : 0.2773  
 3rd Qu.:1.0      3rd Qu.:3220   3rd Qu.:78.749   3rd Qu.: 0.8548  
 Max.   :1.0      Max.   :4293   Max.   :87.822   Max.   :12.2492  
   velocity_Z          sector           track_le

Verificamos que haya la misma cantidad de registros para cada vehiculo.

In [ ]:
registros_por_vehiculo <- telemetria_vueltas_invalidadas %>%
  group_by(carId) %>%
  summarise(cant_registros = n())

registros_por_vehiculo

carId cant_registros
1 ALP   21470         
2 FER   21470         
3 MCL   21470         
4 MER   21470         
5 RED   21470         
6 WIL   21470

In [ ]:
noisy_data_ <- telemetria_vueltas_invalidadas %>%
  group_by(carId, lapNum) %>%
  summarise(
    conteo_total = n(),

    rango_correcto = min(binIndex) == 0 & max(binIndex) == 4293,
    duplicados = conteo_total - n_distinct(binIndex),

    .groups = 'drop'
  )

auditoria_bin_index

carId lapNum conteo_total rango_correcto duplicados
1  ALP   0      4294         TRUE           0         
2  ALP   1      4294         TRUE           0         
3  ALP   2      4294         TRUE           0         
4  ALP   3      4294         TRUE           0         
5  ALP   4      4294         TRUE           0         
6  FER   0      4294         TRUE           0         
7  FER   1      4294         TRUE           0         
8  FER   2      4294         TRUE           0         
9  FER   3      4294         TRUE           0         
10 FER   4      4294         TRUE           0         
11 MCL   0      4294         TRUE           0         
12 MCL   1      4294         TRUE           0         
13 MCL   2      4294         TRUE           0         
14 MCL   3      4294         TRUE           0         
15 MCL   4      4294         TRUE           0         
16 MER   0      4294         TRUE           0         
17 MER   1      4294         TRUE           0         
18 MER   2      4294         TRUE           0         
19 MER   3      4294         TRUE           0         
20 MER   4      4294         TRUE           0         
21 RED   0      4294         TRUE           0         
22 RED   1      4294         TRUE           0         
23 RED   2      4294         TRUE           0         
24 RED   3      4294         TRUE           0         
25 RED   4      4294         TRUE           0         
26 WIL   0      4294         TRUE           0         
27 WIL   1      4294         TRUE           0         
28 WIL   2      4294         TRUE           0         
29 WIL   3      4294         TRUE           0         
30 WIL   4      4294         TRUE           0

carId lapNum Min_Distancia

In [ ]:
test_tiempo_estricto <- telemetria_vueltas_invalidadas %>%
  group_by(carId, lapNum) %>%
  mutate(
    falla_tiempo = lap_time <= lag(lap_time)
  ) %>%
  summarise(
    cant_tiempo_nocreciente = sum(falla_tiempo, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  filter(cant_tiempo_nocreciente > 0)

test_tiempo_estricto

carId lapNum cant_tiempo_nocreciente
1  ALP   0      2                      
2  ALP   1      5                      
3  ALP   2      2                      
4  ALP   3      2                      
5  ALP   4      4                      
6  FER   0      1                      
7  FER   1      4                      
8  FER   2      2                      
9  FER   3      3                      
10 MCL   0      1                      
11 MCL   1      2                      
12 MCL   2      2                      
13 MCL   3      6                      
14 MCL   4      2                      
15 MER   0      7                      
16 MER   2      3                      
17 MER   3      4                      
18 MER   4      1                      
19 RED   0      4                      
20 RED   2      4                      
21 RED   3      7                      
22 WIL   1      5                      
23 WIL   2      2                      
24 WIL   3      4                      
25 WIL   4      3

In [ ]:
options(repr.plot.width = 20, repr.plot.height = 14)